File: mobile/lib/services/socket_service.dart
import 'package:flutter/foundation.dart';
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'auth_service.dart';

class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  SocketService._internal();

  IO.Socket? _socket;
  final _authService = AuthService();
  
  // Callbacks for different events
  Function(Map<String, dynamic>)? onNewRideRequest;
  Function(Map<String, dynamic>)? onRideAccepted;
  Function(Map<String, dynamic>)? onRideStarted;
  Function(Map<String, dynamic>)? onRideCompleted;
  Function(Map<String, dynamic>)? onRideCancelled;
  Function(Map<String, dynamic>)? onDriverLocationUpdated;

  bool get isConnected => _socket?.connected ?? false;

  // Connect to Socket.io server
  Future<void> connect() async {
    if (_socket?.connected ?? false) {
      debugPrint('Socket already connected');
      return;
    }

    try {
      final token = _authService.token;
      if (token == null) {
        debugPrint('No token available for socket connection');
        return;
      }

      _socket = IO.io(
        'http://localhost:4000',
        IO.OptionBuilder()
            .setTransports(['websocket'])
            .enableAutoConnect()
            .setAuth({'token': token})
            .build(),
      );

      _setupEventListeners();
      
      debugPrint('Socket.io connecting...');
    } catch (e) {
      debugPrint('Error connecting to socket: $e');
    }
  }

  // Setup event listeners
  void _setupEventListeners() {
    _socket?.onConnect((_) {
      debugPrint('✅ Socket.io connected');
    });

    _socket?.onDisconnect((_) {
      debugPrint('❌ Socket.io disconnected');
    });

    _socket?.onConnectError((error) {
      debugPrint('Socket connection error: $error');
    });

    _socket?.onError((error) {
      debugPrint('Socket error: $error');
    });

    // Driver events
    _socket?.on('ride:new_request', (data) {
      debugPrint('🚗 New ride request: $data');
      if (onNewRideRequest != null) {
        onNewRideRequest!(Map<String, dynamic>.from(data));
      }
    });

    _socket?.on('ride:cancelled', (data) {
      debugPrint('❌ Ride cancelled: $data');
      if (onRideCancelled != null) {
        onRideCancelled!(Map<String, dynamic>.from(data));
      }
    });

    // Rider events
    _socket?.on('ride:accepted', (data) {
      debugPrint('✅ Ride accepted: $data');
      if (onRideAccepted != null) {
        onRideAccepted!(Map<String, dynamic>.from(data));
      }
    });

    _socket?.on('ride:started', (data) {
      debugPrint('🚀 Ride started: $data');
      if (onRideStarted != null) {
        onRideStarted!(Map<String, dynamic>.from(data));
      }
    });

    _socket?.on('ride:completed', (data) {
      debugPrint('🏁 Ride completed: $data');
      if (onRideCompleted != null) {
        onRideCompleted!(Map<String, dynamic>.from(data));
      }
    });

    _socket?.on('driver:location:updated', (data) {
      debugPrint('📍 Driver location updated: $data');
      if (onDriverLocationUpdated != null) {
        onDriverLocationUpdated!(Map<String, dynamic>.from(data));
      }
    });
  }

  // Emit driver location update
  void emitDriverLocationUpdate({
    required double lat,
    required double lng,
    String? riderId,
    String? eta,
  }) {
    if (!isConnected) {
      debugPrint('Socket not connected, cannot emit location');
      return;
    }

    _socket?.emit('driver:location:update', {
      'location': {'lat': lat, 'lng': lng},
      'riderId': riderId,
      'eta': eta,
    });
  }

  // Emit driver status update
  void emitDriverStatusUpdate(bool isOnline) {
    if (!isConnected) return;
    
    _socket?.emit('driver:status:update', {
      'isOnline': isOnline,
    });
  }

  // Emit ride acceptance
  void emitRideAccept({
    required String rideId,
    required String riderId,
    required Map<String, dynamic> driverInfo,
    required DateTime estimatedArrival,
  }) {
    if (!isConnected) return;

    _socket?.emit('ride:accept', {
      'rideId': rideId,
      'riderId': riderId,
      'driverInfo': driverInfo,
      'estimatedArrival': estimatedArrival.toIso8601String(),
    });
  }

  // Emit ride start
  void emitRideStart({
    required String rideId,
    required String riderId,
  }) {
    if (!isConnected) return;

    _socket?.emit('ride:start', {
      'rideId': rideId,
      'riderId': riderId,
      'startTime': DateTime.now().toIso8601String(),
    });
  }

  // Emit ride completion
  void emitRideComplete({
    required String rideId,
    required String riderId,
    required double fare,
  }) {
    if (!isConnected) return;

    _socket?.emit('ride:complete', {
      'rideId': rideId,
      'riderId': riderId,
      'fare': fare,
      'endTime': DateTime.now().toIso8601String(),
    });
  }

  // Emit ride cancellation
  void emitRideCancel({
    required String rideId,
    String? riderId,
    String? driverId,
    String? reason,
  }) {
    if (!isConnected) return;

    _socket?.emit('ride:cancel', {
      'rideId': rideId,
      'riderId': riderId,
      'driverId': driverId,
      'reason': reason ?? 'Cancelled',
    });
  }

  // Disconnect socket
  void disconnect() {
    _socket?.disconnect();
    _socket?.dispose();
    _socket = null;
    debugPrint('Socket.io disconnected');
  }

  // Dispose
  void dispose() {
    disconnect();
  }
}

File: backend/services/socketService.js
const socketIO = require('socket.io');
const jwt = require('jsonwebtoken');

let io;
const connectedDrivers = new Map(); // Map of driverId -> socketId
const connectedRiders = new Map(); // Map of riderId -> socketId

const initializeSocket = (server) => {
  io = socketIO(server, {
    cors: {
      origin: '*',
      methods: ['GET', 'POST']
    }
  });

  // Authentication middleware
  io.use((socket, next) => {
    try {
      const token = socket.handshake.auth.token;
      
      if (!token) {
        return next(new Error('Authentication error'));
      }

      const decoded = jwt.verify(token, process.env.JWT_SECRET);
      socket.userId = decoded.userId;
      socket.userRole = decoded.role;
      
      next();
    } catch (error) {
      next(new Error('Authentication error'));
    }
  });

  io.on('connection', (socket) => {
    console.log(`User connected: ${socket.userId} (${socket.userRole})`);

    // Store connection based on role
    if (socket.userRole === 'driver') {
      connectedDrivers.set(socket.userId, socket.id);
      socket.join('drivers');
      console.log(`Driver ${socket.userId} joined drivers room`);
    } else if (socket.userRole === 'rider') {
      connectedRiders.set(socket.userId, socket.id);
      socket.join('riders');
      console.log(`Rider ${socket.userId} joined riders room`);
    }

    // Driver location update
    socket.on('driver:location:update', (data) => {
      if (socket.userRole === 'driver') {
        // Broadcast location to specific rider if in active ride
        if (data.riderId) {
          const riderSocketId = connectedRiders.get(data.riderId);
          if (riderSocketId) {
            io.to(riderSocketId).emit('driver:location:updated', {
              driverId: socket.userId,
              location: data.location,
              eta: data.eta
            });
          }
        }
      }
    });

    // Driver online status
    socket.on('driver:status:update', (data) => {
      if (socket.userRole === 'driver') {
        console.log(`Driver ${socket.userId} is now ${data.isOnline ? 'online' : 'offline'}`);
      }
    });

    // Ride request from rider
    socket.on('ride:request', (rideData) => {
      if (socket.userRole === 'rider') {
        // Broadcast to all online drivers in the area
        socket.to('drivers').emit('ride:new_request', {
          rideId: rideData.rideId,
          riderId: socket.userId,
          origin: rideData.origin,
          destination: rideData.destination,
          fare: rideData.fare,
          vehicleType: rideData.vehicleType
        });
        console.log(`New ride request from rider ${socket.userId}`);
      }
    });

    // Driver accepts ride
    socket.on('ride:accept', (data) => {
      if (socket.userRole === 'driver') {
        const riderSocketId = connectedRiders.get(data.riderId);
        if (riderSocketId) {
          io.to(riderSocketId).emit('ride:accepted', {
            rideId: data.rideId,
            driverId: socket.userId,
            driverInfo: data.driverInfo,
            estimatedArrival: data.estimatedArrival
          });
          console.log(`Driver ${socket.userId} accepted ride ${data.rideId}`);
        }
      }
    });

    // Driver starts ride
    socket.on('ride:start', (data) => {
      if (socket.userRole === 'driver') {
        const riderSocketId = connectedRiders.get(data.riderId);
        if (riderSocketId) {
          io.to(riderSocketId).emit('ride:started', {
            rideId: data.rideId,
            driverId: socket.userId,
            startTime: data.startTime
          });
          console.log(`Driver ${socket.userId} started ride ${data.rideId}`);
        }
      }
    });

    // Driver completes ride
    socket.on('ride:complete', (data) => {
      if (socket.userRole === 'driver') {
        const riderSocketId = connectedRiders.get(data.riderId);
        if (riderSocketId) {
          io.to(riderSocketId).emit('ride:completed', {
            rideId: data.rideId,
            driverId: socket.userId,
            fare: data.fare,
            endTime: data.endTime
          });
          console.log(`Driver ${socket.userId} completed ride ${data.rideId}`);
        }
      }
    });

    // Ride cancelled
    socket.on('ride:cancel', (data) => {
      if (socket.userRole === 'rider') {
        const driverSocketId = connectedDrivers.get(data.driverId);
        if (driverSocketId) {
          io.to(driverSocketId).emit('ride:cancelled', {
            rideId: data.rideId,
            riderId: socket.userId,
            reason: data.reason
          });
        }
      } else if (socket.userRole === 'driver') {
        const riderSocketId = connectedRiders.get(data.riderId);
        if (riderSocketId) {
          io.to(riderSocketId).emit('ride:cancelled', {
            rideId: data.rideId,
            driverId: socket.userId,
            reason: data.reason
          });
        }
      }
    });

    // Disconnect
    socket.on('disconnect', () => {
      console.log(`User disconnected: ${socket.userId}`);
      
      if (socket.userRole === 'driver') {
        connectedDrivers.delete(socket.userId);
      } else if (socket.userRole === 'rider') {
        connectedRiders.delete(socket.userId);
      }
    });
  });

  return io;
};

// Helper functions to emit events from controllers
const emitToDriver = (driverId, event, data) => {
  const socketId = connectedDrivers.get(driverId);
  if (socketId && io) {
    io.to(socketId).emit(event, data);
    return true;
  }
  return false;
};

const emitToRider = (riderId, event, data) => {
  const socketId = connectedRiders.get(riderId);
  if (socketId && io) {
    io.to(socketId).emit(event, data);
    return true;
  }
  return false;
};

const emitToAllDrivers = (event, data) => {
  if (io) {
    io.to('drivers').emit(event, data);
    return true;
  }
  return false;
};

const emitToAllRiders = (event, data) => {
  if (io) {
    io.to('riders').emit(event, data);
    return true;
  }
  return false;
};

const getConnectedDrivers = () => {
  return Array.from(connectedDrivers.keys());
};

const getConnectedRiders = () => {
  return Array.from(connectedRiders.keys());
};

const isDriverOnline = (driverId) => {
  return connectedDrivers.has(driverId);
};

const isRiderOnline = (riderId) => {
  return connectedRiders.has(riderId);
};

module.exports = {
  initializeSocket,
  emitToDriver,
  emitToRider,
  emitToAllDrivers,
  emitToAllRiders,
  getConnectedDrivers,
  getConnectedRiders,
  isDriverOnline,
  isRiderOnline
};
